registry/remote/repository.go

```diff
@@ -361,38 +436,86 @@ func (r *Repository) referrers(ctx context.Context, artifactType string, fn func
 	if resp.StatusCode != http.StatusOK {
 		return "", errutil.ParseErrorResponse(resp)
 	}
-	if !legacyAPI {
-		if err := verifyOrasApiVersion(resp); err != nil {
-			return "", err
-		}
-	}
 
-	var page struct {
-		References []ocispec.Descriptor `json:"references"`
-		Referrers  []ocispec.Descriptor `json:"referrers"`
-	}
+	var index ocispec.Index
 	lr := limitReader(resp.Body, r.MaxMetadataBytes)
-	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+	if err := json.NewDecoder(lr).Decode(&index); err != nil {
 		return "", fmt.Errorf("%s %q: failed to decode response: %w", resp.Request.Method, resp.Request.URL, err)
 	}
-	var refs []ocispec.Descriptor
-	if legacyAPI {
-		refs = page.References
-	} else {
-		refs = page.Referrers
+	referrers := index.Manifests
+	if artifactType != "" &&
+		!isReferrersFilterApplied(index.Annotations, "artifactType") {
+		// perform client side filtering if the filter is not applied on the server side
+		referrers = filterReferrers(referrers, artifactType)
 	}
-	// Server may not support filtering. We still need to filter on client side
-	// for sure.
-	refs = filterReferrers(refs, artifactType)
-	if len(refs) > 0 {
-		if err := fn(refs); err != nil {
+	if len(referrers) > 0 {
+		if err := fn(referrers); err != nil {
 			return "", err
 		}
 	}
-
 	return parseLink(resp)
 }
 
+// referrersByTagSchema lists the descriptors of manifests directly referencing
+// the given manifest descriptor by requesting referrers tag.
+// fn is called for the referrers result. If artifactType is not empty,
+// only referrers of the same artifact type are fed to fn.
+// reference: https://github.com/opencontainers/distribution-spec/blob/v1.1.0-rc1/spec.md#backwards-compatibility
+func (r *Repository) referrersByTagSchema(ctx context.Context, desc ocispec.Descriptor, artifactType string, fn func(referrers []ocispec.Descriptor) error) error {
+	ctx = registryutil.WithScopeHint(ctx, r.Reference, auth.ActionPull)
+	referrersTag := buildReferrersTag(desc)
+	desc, rc, err := r.FetchReference(ctx, referrersTag)
+	if err != nil {
+		if errors.Is(err, errdef.ErrNotFound) {
+			// no referrers to the manifest
+			return nil
+		}
+		return err
+	}
+	defer rc.Close()
+
+	var index ocispec.Index
+	lr := limitReader(rc, r.MaxMetadataBytes)
+	indexJSON, err := content.ReadAll(lr, desc)
+	if err != nil {
+		return fmt.Errorf("failed to read referrers index from referrers tag %s: %w", referrersTag, err)
+	}
+	if err := json.Unmarshal(indexJSON, &index); err != nil {
+		return fmt.Errorf("failed to decode referrers index from referrers tag %s: %w", referrersTag, err)
+	}
```

Comment:
We can follow the same logic in the `referrersPageByAPI()`.
```suggestion
	if err := json.NewDecoder(lr).Decode(&index); err != nil {
		return fmt.Errorf("failed to read referrers index from referrers tag %s: %w", referrersTag, err)
	}
```

Reply:
Used `content.VerifyReader`